src/ensembl/src/shared/select/Select.tsx

```diff
@@ -0,0 +1,175 @@
+
+import { splitFromSelected } from './helpers/select-helpers';
+import * as keyCodes from 'src/shared/constants/keyCodes';
+
+import SelectOptionsPanel from './SelectOptionsPanel';
+import SelectArrowhead from './SelectArrowhead';
+
+import styles from './Select.scss';
+
+export type Option = {
+  value: any;
+  label: React.ReactNode;
+  isSelected: boolean;
+  isDisabled?: boolean;
+};
+
+export type OptionGroup = {
+  title?: string;
+  options: Option[];
+};
+
+export type GroupedOptionIndex = [
+  number, // index of option group
+  number // index of the option within a group
+];
+
+type ClosedSelectProps = {
+  isOpen: boolean;
+  selectedOption: Option | null;
+  placeholder: string;
+  onClick: () => void;
+};
+
+type OptionsSpecificProps = {
+  options: Option[];
+  title?: string;
+};
+
+type OptionGroupsSpecificProps = {
+  optionGroups: OptionGroup[];
+};
+
+type CommonProps = {
+  onSelect: (value: any) => void;
+  placeholder?: string;
+};
+
+type OptionsSelectProps = CommonProps & OptionsSpecificProps;
+type OptionGroupsSelectProps = CommonProps & OptionGroupsSpecificProps;
+
+type SelectAdapterProps = OptionsSelectProps | OptionGroupsSelectProps;
+
+type SelectProps = OptionGroupsSelectProps & { placeholder: string };
+
+// TODO: think of a better name for this component?
+const ClosedSelect = (props: ClosedSelectProps) => {
+  const className = props.isOpen
+    ? styles.selectClosedInvisible
+    : styles.selectClosed;
+  return (
+    <span className={className} onClick={props.onClick}>
+      {props.selectedOption ? props.selectedOption.label : props.placeholder}
+      <SelectArrowhead />
+    </span>
+  );
+};
+
+const Select = (props: SelectProps) => {
+  const [isOpen, setIsOpen] = useState(false);
+  const [selectedOption, optionGroups] = splitFromSelected(props.optionGroups);
+
+  // use ref to keep track of whether the element is focused
+  // (using ref makes this value available to the callback called from useEffect)
+  const focusRef = useRef(false);
+
+  useEffect(() => {
+    document.addEventListener('keydown', handleKeyPress);
+    return () => document.removeEventListener('keyup', handleKeyPress);
+  }, []);
+
+  const openPanel = () => {
+    setIsOpen(true);
+  };
+
+  const closePanel = () => {
+    setIsOpen(false);
+  };
+
+  const handleFocus = () => {
+    focusRef.current = true;
+  };
+
+  const handleBlur = () => {
+    focusRef.current = false;
+    if (isOpen) {
+      setIsOpen(false);
+    }
+  };
+
+  const handleKeyPress = (event: KeyboardEvent) => {
+    if (
+      !(
+        focusRef.current &&
+        [keyCodes.ENTER, keyCodes.ESC].includes(event.keyCode)
+      )
+    ) {
+      return;
+    }
+
+    if (event.keyCode === keyCodes.ENTER) {
+      setIsOpen(true);
+    } else if (event.keyCode === keyCodes.ESC) {
+      setIsOpen(false);
+    }
+  };
+
+  const handleSelect = (optionIndex: GroupedOptionIndex) => {
+    const [groupdIndex, itemIndex] = optionIndex;
```

Comment:
Is `groupdIndex` intentionally named like that or is it a typo?

Reply:
That's a typo; should be groupIndex.